### PR TITLE
feat(foundationdb): introduce db_run

### PR DESCRIPTION
**Pull request by PierreZ**
The idea is to add the `db.run` method, which is automatically handling Transaction retry. 

This is a second try(#54 and #60), and while I know it is not perfect, it provides a great developer experience. This method is mandatory to build effective layers with the bindings, and we are currently using it at Clever Cloud. 

It is also a requirement for #70 and to open-source the [simulation crate](https://forums.foundationdb.org/t/cpp-implementation-of-an-externalworkload/3380/6) that we have internally.

A new type, `RetryableTransaction`, has been added, which is almost like a Transaction, except that:
* it does not provide [cancel](),
* it is cloneable.

Users can still use the `transact`/`transact_boxed` methods, if they want to collaborate more with the borrow-checker, at the price of a more difficult API.

Remaining things to do if this gets merged:

* convert examples to `db.run`,
* find a way to use the borrow-checker to avoid the `Arc::try_unwrap` during `on_error` methods.

### CODE REVIEW

File: foundationdb/src/transaction.rs [hunk 1]
```diff
@@ -1018,3 +1021,226 @@ impl Drop for Transaction {
...
+        self.inner.get_ranges_keyvalues(opt, snapshot)
+    }
+
+    /// Implements the recommended retry and backoff behavior for a transaction. This function knows
+    /// which of the error codes generated by other `Transaction` functions represent temporary
+    /// error conditions and which represent application errors that should be handled by the
+    /// application. It also implements an exponential backoff strategy to avoid swamping the
+    /// database cluster with excessive retries when there is a high level of conflict between
+    /// transactions.
+    pub async fn on_error(self, err: FdbError) -> Result<RetryableTransaction, FdbError> {
```
Comment (Speedy37) [inline foundationdb/src/transaction.rs:1071]: document panic behavior

File: foundationdb/src/transaction.rs [hunk 2]
```diff
@@ -1018,3 +1021,226 @@ impl Drop for Transaction {
...
+
+    /// Attempts to commit the sets and clears previously applied to the database snapshot
+    /// represented by transaction to the actual database.
+    ///
+    /// The commit may or may not succeed – in particular, if a conflicting transaction previously
+    /// committed, then the commit must fail in order to preserve transactional isolation. If the
+    /// commit does succeed, the transaction is durably committed to the database and all
+    /// subsequently started transactions will observe its effects.
+    pub async fn commit(self) -> Result<TransactionCommitted, TransactionCommitError> {
+        match Arc::try_unwrap(self.inner) {
```
Comment (Speedy37) [inline foundationdb/src/transaction.rs:1104]: document panic behavior

File: foundationdb/src/database.rs [hunk 3]
```diff
@@ -109,6 +109,17 @@ impl Database {
...
 
+    fn create_retryable_trx(&self) -> FdbResult<RetryableTransaction> {
+        let mut trx: *mut fdb_sys::FDBTransaction = std::ptr::null_mut();
+        let err =
+            unsafe { fdb_sys::fdb_database_create_transaction(self.inner.as_ptr(), &mut trx) };
+        error::eval(err)?;
+        Ok(RetryableTransaction::new(Transaction::new(
+            NonNull::new(trx)
+                .expect("fdb_database_create_transaction to not return null if there is no error"),
+        )))
```
Comment (Speedy37) [inline foundationdb/src/database.rs:120]:
```suggestion
        Ok(RetryableTransaction::new(self.create_trx()?))
```

File: foundationdb/src/database.rs [hunk 4]
```diff
@@ -218,6 +229,76 @@ impl Database {
         )
     }
 
+    /// Runs a transactional function against this Database with retry logic.
+    /// The associated closure will be called until a non-retryable FDBError
+    /// is thrown or commit(), returns success.
+    /// This call is blocking -- this method will not return until commit() has been called and returned success.
```
Comment (Speedy37) [inline foundationdb/src/database.rs:235]: well it's async, so not really blocking :)

**Review by Speedy37 — APPROVED** [review]
looks good, it's a nice tradeoff

**Comment by Speedy37** [comment]
About the panic behavior.
While I agree, misusing RetryableTransaction is a bug and should, it might be good to return an Error to not crash production  (maybe only in release build? or via a feature).

GAT is around the corner (next stable release) so I might give a shot at writing a more user friendly transact method but that would not change the fact that it's a &Transaction that is manipulated inside the closure.

I wonder, what was the pain of manipulating &Transaction, cause I remember having a whole object db engine (multithreaded) inside the closure and that was not a big problem. Still a few `struct Foo<'trx> { }` there and there, but forced corrected was what I wanted.

**Comment by Speedy37** [comment]
About that point: "find a way to use the borrow-checker to avoid the Arc::try_unwrap during on_error methods."

No lifetime, no way at compile time to prove you are the only remaining owner of the transaction.

File: foundationdb/src/transaction.rs (hunk 1 above)
Comment (PierreZ) [reply foundationdb/src/transaction.rs:1071]: Done

File: foundationdb/src/transaction.rs (hunk 2 above)
Comment (PierreZ) [reply foundationdb/src/transaction.rs:1104]: Done

File: foundationdb/src/database.rs (hunk 3 above)
Comment (PierreZ) [reply foundationdb/src/database.rs:120]: You are right, much easier :tongue:

File: foundationdb/src/database.rs (hunk 4 above)
Comment (PierreZ) [reply foundationdb/src/database.rs:235]: This is misleading indeed, changing

**Comment by PierreZ** [comment]
> About the panic behavior.
> While I agree, misusing RetryableTransaction is a bug and should, it might be good to return an Error to not crash production (maybe only in release build? or via a feature).

I am reworking the error from db-run to include all errors that can be thrown

**Comment by PierreZ** [comment]
What do you think @Speedy37 about the new error?

**Comment by PierreZ** [comment]
Forgot a panic, will push another commit

File: foundationdb/src/error.rs [hunk 5]
```diff
@@ -74,11 +78,87 @@ impl FdbError {
...
+                    None
+                }
+            }
+            FdbBindingError::HcaError(hca_error) => {
+                if let HcaError::FdbError(e) = hca_error {
+                    Some(*e)
+                } else {
+                    None
+                }
+            }
```
Comment (Speedy37) [inline foundationdb/src/error.rs:121]:
```suggestion
            FdbBindingError::DirectoryError(DirectoryError::FdbError(e)) = > Some(*e),
            FdbBindingError::HcaError(HcaError::FdbError(e)) = > Some(*e),
```

**Review by Speedy37 — APPROVED** [review]

**Comment by Speedy37** [comment]
> What do you think @Speedy37 about the new error?

Looks ok to me.

**Comment by Speedy37** [comment]
I just though that `Arc<Transaction>` (via `Deref`) already kinda exposed the same API as `RetryableTransaction `does (ie. everything that takes `&self`).

`RetryableTransaction::commit/on_error` are non public making `RetryableTransaction` almost the same as `Arc<Transaction>`

Another impl for RetryableTransaction  could be:
```rust
#[derive(Clone)]
pub struct RetryableTransaction {
    inner: Arc<Transaction>,
}
impl std::ops::Deref for RetryableTransaction  {
    type Target = Transaction;
    fn deref(&self) -> &Transaction {
        self.inner.deref()
    }
}
impl RetryableTransaction {
    pub(crate) async fn on_error(
        self,
        err: FdbError,
    ) -> Result<Result<RetryableTransaction, FdbError>, FdbBindingError> {
        // checking weak references
        if Arc::weak_count(&self.inner) != 0 {
            return Err(FdbBindingError::ReferenceToTransactionKept);
        }

        // checking strong references
        match Arc::try_unwrap(self.inner) {
            Ok(trx) => match trx.on_error(err).await {
                Ok(trx) => Ok(Ok(RetryableTransaction::new(trx))),
                Err(err) => Ok(Err(err)),
            },
            Err(_) => Err(FdbBindingError::ReferenceToTransactionKept),
        }
    }
    pub(crate) async fn commit(
        self,
    ) -> Result<Result<TransactionCommitted, TransactionCommitError>, FdbBindingError> {
        // check weak references
        if Arc::weak_count(&self.inner) != 0 {
            return Err(FdbBindingError::ReferenceToTransactionKept);
        }
        match Arc::try_unwrap(self.inner) {
            Ok(inner) => Ok(inner.commit().await),
            Err(_) => Err(FdbBindingError::ReferenceToTransactionKept),
        }
    }
}
```

**Comment by PierreZ** [comment]
> I just though that `Arc<Transaction>` (via `Deref`) already kinda exposed the same API as `RetryableTransaction `does (ie. everything that takes `&self`).
> 
> `RetryableTransaction::commit/on_error` are non public making `RetryableTransaction` almost the same as `Arc<Transaction>`
> 
> Another impl for RetryableTransaction could be:
> 
> ```rust
> #[derive(Clone)]
> pub struct RetryableTransaction {
>     inner: Arc<Transaction>,
> }
> impl std::ops::Deref for RetryableTransaction {
>     type Target = Transaction;
>     fn deref(&self) -> &Transaction {
>         self.inner.deref()
>     }
> }
> impl RetryableTransaction {
>     pub(crate) fn new(t: Transaction) -> RetryableTransaction {
>         RetryableTransaction { inner: Arc::new(t) }
>     }
>     pub(crate) fn take(self) -> Result<Transaction, FdbBindingError> {
>         // checking weak references
>         if Arc::weak_count(&self.inner) != 0 {
>             return Err(FdbBindingError::ReferenceToTransactionKept);
>         }
>         Arc::try_unwrap(self.inner).map_err(|_| FdbBindingError::ReferenceToTransactionKept)
>     }
>     pub(crate) async fn on_error(
>         self,
>         err: FdbError,
>     ) -> Result<Result<RetryableTransaction, FdbError>, FdbBindingError> {
>         Ok(self
>             .take()?
>             .on_error(err)
>             .await
>             .map(RetryableTransaction::new))
>     }
>     pub(crate) async fn commit(
>         self,
>     ) -> Result<Result<TransactionCommitted, TransactionCommitError>, FdbBindingError> {
>         self.take()?.commit().await
>     }
> }
> ```

This is brilliant :exploding_head: This is making the code much more readable, thanks a lot @Speedy37 :pray: 

I think you deserve all the credits here, would you like to push a commit to my branch? Or may I include it in a commit?

**Comment by PierreZ** [comment]
I integrated your proposal. Feel free to merge it @Speedy37 if you're ok with this.

**Pull request merged**